### DISCUSSION TITLE
Systemwide key handler

### DISCUSSION THREAD

**Discussion by markusamshove** [AvaloniaUI/Avalonia#13368] [Q&A]
Hello,

I'm currently prototyping to port a WPF application to Avalonia.

In the WPF application I'm using [NHotkey](https://github.com/thomaslevesque/NHotkey/) to register a global keyhandler to open my application (think of it like an Applauncher).

Is this something Avalonia could support itself, or is there already some platform independent implementation in Avalonia that I could reuse?

**Comment by DylanSnel** [comment]
Did you ever get an answer?

**Comment by timunie** [comment]
System wide keys need to be implemented by the App. There is no x-plat way to do it properly afaik.

**Reply by thevortexcloud** [reply]
It's also fairly complicated to do too as every platform has their own way of handling keys and doing key codes. There is this library though that supposedly does it for you.

https://github.com/TolikPylypchuk/SharpHook

But I have never tried it and it looks like it has a dependency on X11 on Linux (which is slowly being removed from most major distros).

On Linux you can somewhat easily make a global key handler that is not dependent on X11 by parsing the key input file under `/dev/input`and using a `FileSystemWatcher`. But it's a binary file so you also need to decode it yourself. It also requires root to read/custom udev rules.

**Reply by DylanSnel** [reply]
Any good resources on platform specific code?

**Reply by thevortexcloud** [reply]
On Windows it requires going into native code. Apparently there is a C hook you can use 

https://stackoverflow.com/questions/9149600/global-keyboard-hooks-in-c

That said I have only seen it implemented using a background thread polling [`GetKeyState`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getkeystate) for specific keys that need to be watched.

On Linux if you want to decode that file I mentioned, you can do that entirely from within C#. There is some basic information here on how to decode the file here: 

https://superuser.com/questions/1310494/how-to-decode-logged-dev-input-data

Do note that `timeval` is not a fixed width on Linux. It can vary based on how the system has been compiled. It's _generally_ 8 or 16 bytes, for 32 bit and 64 bit systems respectively though. But it can really be anything.

If you want to use X11 APIs then that library I linked before already has most of the logic you need for that.

As for Wayland as far as I am aware there is no way to do this yet on Wayland.


I am not too sure about any other platforms. But all platforms you will need to map the keys from the native keycode to some generic key code so your app does not care what platform it is on. Windows has [these virtual keys](https://learn.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes) and Linux [has these codes.](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h)

**Comment by timunie** [comment]
as this is more a question than an actual bug report, I'll make a discussion from this issue for now.
